exo/inference/pytorch/inference.py

```diff
@@ -0,0 +1,301 @@
+# experimental, based off of tinygrad/inference.py
+import os
+import re
+import numpy as np
+import torch
+import json
+
+from typing import Optional, Tuple
+from exo.inference.shard import Shard
+from exo.inference.inference_engine import InferenceEngine
+from exo.inference.pytorch.model.hf import ShardedHuggingFaceModel
+from exo.inference.tokenizers import resolve_tokenizer
+from exo.helpers import DEBUG
+from exo.download.hf.hf_shard_download import HFShardDownloader
+
+from transformers import AutoTokenizer
+# llama
+from transformers.models.llama.modeling_llama import LlamaModel
+
+# model value options
+TOP_K = 20
+TEMP = 0.6
+TOP_P = 0.9
+MAX_LENGTH = 125
+MAX_TIME = 60.0
+
+class PyTorchDynamicShardInferenceEngine(InferenceEngine):
+  """
+  PyTorch Dynamic Shard Inference Engine for performing model inference with sharded Pytorch/HF based models.
+  """
+
+  def __init__(self, shard_downloader: HFShardDownloader):
+    """
+    Initialize the inference engine.
+
+    Args:
+      debug (bool): If True, enables debug logging. Defaults to False.
+    """
+    self.shard = None
+    self.shard_downloader = shard_downloader
+    self.stateful_sharded_model = None
+    self.tokenizer = None
+
+    # the whole history with new logits need to
+    # be passed to the model to reach the end token
+    # even with caching
+    self.past_input_ids = None
+
+    # setup cuda device
+    if os.environ.get("PYTORCH_DEVICE"):
+      pytorch_device = os.environ["PYTOCH_DEVICE"]
+      if pytorch_device not in ["cuda", "mps", "cpu"]:
```

Comment:
nitpick, but `["cuda", "mps", "cpu"]` can be `["cuda", "mps"]` since you can just leave it as `cpu`